less_users/views.py

```diff
@@ -40,104 +48,231 @@ class ProfileView(LoginRequiredMixin, View):
+                            )
+                        else:
+                            messages.warning(
+                                request,
+                                f"You cannot get new points for this challenge yet: {user_challenge.challenge}. "
+                                f"You have already done it within last {user_challenge.challenge.frequency} days",
+                            )
+                    else:
+                        # try to check previous logs, even for other user_challenges, but for the same user and challenge
+                        # and even if they are not active or are deleted, it counts points!
+                        prev_logs = Log.objects.filter(
+                            user_challenge__user_id=request.user
+                        ).filter(user_challenge__challenge_id=challenge_id)
+                        if prev_logs:
+                            last_log = prev_logs.latest("date").date.date()
+                            if (
+                                last_log
+                                + timedelta(
+                                    days=Challenge.objects.get(
+                                        pk=challenge_id
+                                    ).frequency
+                                )
+                                > date.today()
+                            ):
+                                messages.warning(
+                                    request,
+                                    f"You cannot get new points for this challenge yet: {user_challenge.challenge}. "
+                                    f"You have already done it within last {user_challenge.challenge.frequency} days",
+                                )
+                            else:
+                                new_points = user_challenge.get_points
+                                Log.objects.create(
+                                    user_challenge_id=user_challenge.id, points=points
+                                )
+                                my_profile = request.user.profile
+                                my_profile.points += points
+                                my_profile.save()
+                                messages.success(
+                                    request,
+                                    f"You have just got {new_points} points for challenge: {user_challenge.challenge}!",
+                                )
+                        else:
+                            messages.success(
+                                request,
+                                "Is it possible to reach this if/else part?!",
+                            )
+                else:
+                    """stop user challenge, it is out of date"""
+                    user_challenge.is_active = False
+                    user_challenge.is_deleted = True
+                    user_challenge.save()
```

Comment:
nit: wrzuciłabym do jakieś zewnętrznej metody, np `deactivate`
```python
class UserChallenge:
   def deactivate(self, save=False):
       self.is_active = False
       self.is_deleted = True
       if save:
            self.save()
```

Reply:
hm, podobne działanie mam w widoku `def event_view_delete(request)`... gdybym napisała metodę, jak proponujesz, mogłabym jej wywołaniem zastąpić kod tej funkcji w widoku, prawda?

Reply:
Tak.

Reply:
parametr `save=True` przekazuję gdy chcę wykonać tę metodę: `user_challenge.deactivate(True)`?
A Twoja propozycja funkcji uruchamia mechanizm zapisu tych zmian bezpośrednio wywołaniu, bez potrzeby zewnętrznego zapisywania zmian? Dziękuję, to cenna informacja. Pewnie bym na to nie wpadła i kombinowała zapisywanie zmien pod wywołaną na obiekcie metodą, a tak- jest czysto.